flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java

```diff
@@ -0,0 +1,210 @@
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.annotation.JsonSerialize;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static org.apache.flink.util.StringUtils.isNullOrWhitespaceOnly;
+
+/**
+ * An implementation of the {@link JobResultStore} which persists job result data to an underlying
+ * distributed filesystem.
+ */
+public class FileSystemJobResultStore extends AbstractThreadsafeJobResultStore {
+
+    private static final String DIRTY_SUFFIX = "_DIRTY";
+
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    private final FileSystem fileSystem;
+
+    private final Path basePath;
+
+    private final boolean deleteOnCommit;
+
+    @VisibleForTesting
+    FileSystemJobResultStore(FileSystem fileSystem, Path basePath, boolean deleteOnCommit)
+            throws IOException {
+        this.fileSystem = fileSystem;
+        this.basePath = basePath;
+        this.deleteOnCommit = deleteOnCommit;
+
+        this.fileSystem.mkdirs(this.basePath);
+    }
+
+    public static FileSystemJobResultStore fromConfiguration(Configuration config)
+            throws IOException {
+        Preconditions.checkNotNull(config);
+        final String jrsStoragePath = config.get(JobResultStoreOptions.STORAGE_PATH);
+        final Path basePath;
+
+        if (isNullOrWhitespaceOnly(jrsStoragePath)) {
+            final String haStoragePath = config.get(HighAvailabilityOptions.HA_STORAGE_PATH);
+            final String haClusterId = config.get(HighAvailabilityOptions.HA_CLUSTER_ID);
+            basePath = new Path(haStoragePath + "/job-result-store/" + haClusterId);
+        } else {
+            basePath = new Path(jrsStoragePath);
+        }
+
+        boolean deleteOnCommit = config.get(JobResultStoreOptions.DELETE_ON_COMMIT);
+
+        return new FileSystemJobResultStore(basePath.getFileSystem(), basePath, deleteOnCommit);
+    }
+
+    /**
+     * Given a job ID, construct the path for a dirty entry corresponding to it in the job result
+     * store.
+     *
+     * @param jobId The job ID to construct a dirty entry path from.
+     * @return A path for a dirty entry for the given the Job ID.
+     */
+    private Path constructDirtyPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + DIRTY_SUFFIX + ".json");
+    }
+
+    /**
+     * Given a job ID, construct the path for a clean entry corresponding to it in the job result
+     * store.
+     *
+     * @param jobId The job ID to construct a clean entry path from.
+     * @return A path for a clean entry for the given the Job ID.
+     */
+    private Path constructCleanPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + ".json");
```

Comment:
```suggestion
        return new Path(this.basePath.getPath(), jobId.toString() + ".json");
```
nit: this omits the file separator which is used by `Path`.